File: LeetCode/0202. Happy Number/solution.py
"""
402 / 402 test cases passed.
Runtime: 32 ms
Memory Usage: 14.9 MB
"""
class Solution:
    def isHappy(self, n: int) -> bool:
        def get_next(t):
            nxt = 0
            while t > 0:
                t, digit = divmod(t, 10)
                nxt += digit ** 2
            return nxt
        seen = set()
        while n != 1 and n not in seen:
            seen.add(n)
            n = get_next(n)
        return n == 1

File: LeetCode/0202. Happy Number/solution.cpp
/**
 * 402 / 402 test cases passed.
 * Runtime: 4 ms
 * Memory Usage: 6.2 MB 
 */
class Solution {
public:
    bool isHappy(int n) {
        unordered_set<int> vis;
        vis.emplace(n);
        while (n != 1) {
            int t = 0;
            while (n > 0) {
                t += (n % 10) * (n % 10);
                n /= 10;
            }
            if (vis.count(t)) return false;
            n = t;
            vis.emplace(n);
        }
        return true;
    }
};
